src/backend/mpi/write_buffer.hpp

```diff
@@ -170,16 +188,135 @@ class write_buffer
 					storepageDIFF(cache_index+i,page_size*i+page_address);
 				}
+				// Close any windows used to write back data
+				// This should be replaced with an API call
+				for(int i = 0; i < argo::backend::number_of_nodes(); i++){
+					if(barwindowsused[i] == 1){
+						MPI_Win_unlock(i, globalDataWindow[i]);
+						barwindowsused[i] = 0;
+					}
+				}
+			}
+			double t_end = MPI_Wtime();
+
+			// Update timer statistics
+			std::lock_guard<std::mutex> stat_lock(_stat_mutex);
+			_write_back_time += t_end - t_start;
+			_partial_flush_count++;
+		}
+
+		/**
+		 * @brief	Helper function to call _add
+		 * @param	buf Pointer to a write_buffer object
+		 * @param	val Value to add to buf
+		 */
+		static void _add_helper(write_buffer* buf, T val) {
+			buf->_add(val);
+		}
+
+		/**
+		 * @brief	Internal function to add an element to the write buffer
+		 * @param	val The value of type T to add to the buffer
+		 */
+		void _add(T val) {
+			// If already present in the buffer, do nothing
+			if(has(val)){
+				return;
+			}
+
+			// If the buffer is full, write back _write_back_size indices
+			if(size() >= _max_size){
+				flush_partial();
 			}
 
-			// Close any windows used to write back data
-			// This should be replaced with an API call
-			for(int i = 0; i < argo::backend::number_of_nodes(); i++){
-				if(barwindowsused[i] == 1){
-					MPI_Win_unlock(i, globalDataWindow[i]);
-					barwindowsused[i] = 0;
+			// Add val to the back of the buffer
+			emplace_back(val);
+			std::lock_guard<std::mutex> stat_lock(_stat_mutex);
+			_page_count++;
+		}
+
+		/**
+		 * @brief	Helper function to call _erase
+		 * @param	buf Pointer to a write_buffer object
+		 * @param	val Value to erase from buf
+		 */
+		static void _erase_helper(write_buffer* buf, T val) {
+			buf->_erase(val);
+		}
+
+		/**
+		 * @brief	Internal implementation of erase
+		 * @param	val The value of type T to erase
+		 * @pre		_qd_lock must be taken
+		 */
+		void _erase(T val) {
+			// Attempt to get iterator to element equal to val
+			typename std::deque<T>::iterator it =
+				std::find(_buffer.begin(),_buffer.end(), val);
+			// If found, erase it
+			if(it != _buffer.end()){
+				_buffer.erase(it);
+			}
+		}
+
+		/**
+		 * @brief	Helper function to call _flush
+		 * @param	buf Pointer to a write_buffer object
+		 * @param	flag Flag to signal when _flush is done
+		 */
+		static void _flush_helper(write_buffer* buf, std::atomic<bool>* flag) {
+			buf->_flush(flag);
+		}
+
+		/**
+		 * @brief	Internal implementation of flush
+		 * @param	w_flag Flag to signal when flush is done
+		 * @pre		_qd_lock must be taken
+		 */
+		void _flush(std::atomic<bool>* w_flag) {
+			double t_start = MPI_Wtime();
+
+			// If it's empty we don't need to do anything
+			if(empty()){
+				double t_stop = MPI_Wtime();
+				// Signal that flush is done
+				w_flag->store(true, std::memory_order_release);
+				std::lock_guard<std::mutex> stat_lock(_stat_mutex);
+				_flush_time += t_stop-t_start;
```

Comment:
code duplication with end of function. Please find a way to not repeat yourself.

Reply:
I think we can leave it to the compiler to optimize away the double not-empty check.